scipy/optimize/_root_scalar.py

```diff
@@ -0,0 +1,237 @@
+"""
+Unified interfaces to root finding algorithms for real or complex
+scalar functions.
+
+Functions
+---------
+- root : find a root of a scalar function.
+"""
+from __future__ import division, print_function, absolute_import
+
+import numpy as np
+from scipy._lib.six import callable
+
+from . import zeros as optzeros
+
+__all__ = ['root_scalar']
+
+
+class MemoizeDer(object):
+    """Decorator that caches the value and derivative(s) of function each
+    time it is called.
+
+    This is a simplistic memoizer that calls and caches a single value
+    of `f(x, *args)`.
+    It assumes that `args` does not change between invocations.
+    It supports the use case of a root-finder where `args` is fixed,
+    `x` changes, and only rarely, if at all, does x assume the same value
+    more than once."""
+    def __init__(self, fun):
+        self.fun = fun
+        self.vals = None
+        self.x = None
+        self.n_calls = 0
+
+    def __call__(self, x, *args):
+        r"""Calculate f or used cached value if available"""
```

Comment:
"use" - present tense?

ditto for `fprime` and `fprime2` below

Reply:
Agree.